File: packages/sdk/core/src/scrubbing.ts
/**
 * PII scrubbing — recursively walks event payloads and masks values whose
 * keys or string contents match known sensitive patterns. Pure, side-effect
 * free, returns a new object (deep-cloned) so original references stay intact.
 */
import type { ErrorEvent, SanitizeOptions } from "./types";

const REDACTED = "[Filtered]";

/**
 * Default sensitive key substrings (case-insensitive).
 * Matches if a key *contains* any of these.
 */
const DEFAULT_SENSITIVE_KEYS: readonly string[] = [
  "password",
  "passwd",
  "pwd",
  "secret",
  "token",
  "api_key",
  "apikey",
  "access_token",
  "refresh_token",
  "bearer",
  "authorization",
  "cookie",
  "set-cookie",
  "session",
  "credit_card",
  "creditcard",
  "cc_number",
  "card_number",
  "cvv",
  "cvc",
  "ssn",
  "social_security",
  "private_key",
  "client_secret",
  "signature",
];

/**
 * Default value patterns matched against string values.
 * Matches anywhere in the string. False positives prefer leaving data alone —
 * we only redact full matches we're confident about.
 */
const DEFAULT_VALUE_PATTERNS: readonly RegExp[] = [
  // JWT (3 base64url segments)
  /\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b/,
  // AWS access key
  /\bAKIA[0-9A-Z]{16}\b/,
  // US SSN dashed
  /\b\d{3}-\d{2}-\d{4}\b/,
  // Stripe-style secret keys (sk_live_..., sk_test_...)
  /\bsk_(?:live|test)_[A-Za-z0-9]{16,}\b/,
  // Slack tokens (xox*)
  /\bxox[abprs]-[A-Za-z0-9-]{10,}\b/,
  // GitHub PAT
  /\bghp_[A-Za-z0-9]{36}\b/,
];

const EMAIL_PATTERN = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/;
const CREDIT_CARD_PATTERN = /\b(?:\d[ -]*?){13,19}\b/;

interface CompiledOptions {
  keySet: string[];
  patterns: RegExp[];
  scrubEmails: boolean;
}

function compileOptions(opts: SanitizeOptions | undefined): CompiledOptions {
  const userKeys = opts?.sensitiveKeys || [];
  const keySet = [...DEFAULT_SENSITIVE_KEYS, ...userKeys].map((k) =>
    k.toLowerCase()
  );
  const patterns: RegExp[] = [
    ...DEFAULT_VALUE_PATTERNS,
    ...(opts?.customPatterns || []),
  ];
  return {
    keySet,
    patterns,
    scrubEmails: opts?.scrubEmails === true,
  };
}

function keyIsSensitive(key: string, keySet: string[]): boolean {
  const lower = key.toLowerCase();
  for (const needle of keySet) {
    if (lower.includes(needle)) return true;
  }
  return false;
}

/**
 * Luhn check — used to reduce credit-card false positives on numeric runs.
 * Strips spaces/dashes first, then runs the standard Luhn algorithm.
 */
function luhnCheck(s: string): boolean {
  const digits = s.replace(/[^\d]/g, "");
  if (digits.length < 13 || digits.length > 19) return false;
  let sum = 0;
  let alt = false;
  for (let i = digits.length - 1; i >= 0; i--) {
    let n = parseInt(digits[i]!, 10);
    if (alt) {
      n *= 2;
      if (n > 9) n -= 9;
    }
    sum += n;
    alt = !alt;
  }
  return sum % 10 === 0;
}

function valueIsSensitive(value: string, opts: CompiledOptions): boolean {
  for (const re of opts.patterns) {
    if (re.test(value)) return true;
  }
  if (opts.scrubEmails && EMAIL_PATTERN.test(value)) return true;
  // Credit cards: only flag if Luhn-valid to avoid masking timestamps, IDs, etc.
  if (CREDIT_CARD_PATTERN.test(value) && luhnCheck(value)) return true;
  return false;
}

/**
 * Recursive walker. Replaces sensitive values with "[Filtered]". Preserves
 * shape — never deletes keys. Handles cycles.
 */
function scrubValue(
  value: unknown,
  opts: CompiledOptions,
  seen: WeakSet<object>,
  depth: number
): unknown {
  if (depth > 8) return value;
  if (value === null || value === undefined) return value;

  if (typeof value === "string") {
    return valueIsSensitive(value, opts) ? REDACTED : value;
  }

  if (typeof value !== "object") return value;

  if (seen.has(value as object)) return value;
  seen.add(value as object);

  if (Array.isArray(value)) {
    return value.map((v) => scrubValue(v, opts, seen, depth + 1));
  }

  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
    if (keyIsSensitive(k, opts.keySet)) {
      out[k] = REDACTED;
    } else {
      out[k] = scrubValue(v, opts, seen, depth + 1);
    }
  }
  return out;
}

/**
 * Scrub an entire ErrorEvent in place-safe way (returns a new event,
 * leaves the input untouched). Pass `enabled: false` to short-circuit.
 */
export function scrubEvent(
  event: ErrorEvent,
  options: SanitizeOptions | undefined
): ErrorEvent {
  if (options?.enabled === false) return event;
  const compiled = compileOptions(options);
  const seen = new WeakSet<object>();

  return {
    ...event,
    extra: event.extra
      ? (scrubValue(event.extra, compiled, seen, 0) as Record<string, unknown>)
      : event.extra,
    user: event.user
      ? // user.id and user.email get special treatment — id stays, email is
        // optional based on scrubEmails flag (treated as a value scrub)
        (scrubValue(event.user, compiled, seen, 0) as ErrorEvent["user"])
      : event.user,
    tags: event.tags
      ? (scrubValue(event.tags, compiled, seen, 0) as Record<string, string>)
      : event.tags,
    request: event.request
      ? {
          ...event.request,
          headers: event.request.headers
            ? (scrubValue(
                event.request.headers,
                compiled,
                seen,
                0
              ) as Record<string, string>)
            : event.request.headers,
          data: event.request.data
            ? scrubValue(event.request.data, compiled, seen, 0)
            : event.request.data,
        }
      : event.request,
    breadcrumbs: event.breadcrumbs
      ? event.breadcrumbs.map((b) => ({
          ...b,
          data: b.data
            ? (scrubValue(b.data, compiled, seen, 0) as Record<string, unknown>)
            : b.data,
        }))
      : event.breadcrumbs,
  };
}

// Re-export internals for testing
export const __test = {
  keyIsSensitive,
  valueIsSensitive,
  luhnCheck,
  compileOptions,
  scrubValue,
  DEFAULT_SENSITIVE_KEYS,
};

File: packages/sdk/python/bugwatch/scrubbing.py
"""
PII scrubbing — recursively walks event dicts and masks values whose
keys or string contents match known sensitive patterns.

Returns a new dict (deep-copied). Original references stay intact.
"""

import copy
import re
from re import Pattern
from typing import Any, Optional

REDACTED = "[Filtered]"

# Default sensitive key substrings (case-insensitive).
DEFAULT_SENSITIVE_KEYS: list[str] = [
    "password",
    "passwd",
    "pwd",
    "secret",
    "token",
    "api_key",
    "apikey",
    "access_token",
    "refresh_token",
    "bearer",
    "authorization",
    "cookie",
    "set-cookie",
    "session",
    "credit_card",
    "creditcard",
    "cc_number",
    "card_number",
    "cvv",
    "cvc",
    "ssn",
    "social_security",
    "private_key",
    "client_secret",
    "signature",
]

# Default value patterns matched against string values.
_JWT_RE = re.compile(r"\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b")
_AWS_KEY_RE = re.compile(r"\bAKIA[0-9A-Z]{16}\b")
_SSN_RE = re.compile(r"\b\d{3}-\d{2}-\d{4}\b")
_STRIPE_RE = re.compile(r"\bsk_(?:live|test)_[A-Za-z0-9]{16,}\b")
_SLACK_RE = re.compile(r"\bxox[abprs]-[A-Za-z0-9-]{10,}\b")
_GITHUB_PAT_RE = re.compile(r"\bghp_[A-Za-z0-9]{36}\b")
_EMAIL_RE = re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b")
_CC_RE = re.compile(r"\b(?:\d[ -]*?){13,19}\b")

DEFAULT_VALUE_PATTERNS: list[Pattern[str]] = [
    _JWT_RE,
    _AWS_KEY_RE,
    _SSN_RE,
    _STRIPE_RE,
    _SLACK_RE,
    _GITHUB_PAT_RE,
]


def _luhn_check(s: str) -> bool:
    """Luhn algorithm to reduce credit-card false positives."""
    digits = re.sub(r"[^\d]", "", s)
    if len(digits) < 13 or len(digits) > 19:
        return False
    total = 0
    alt = False
    for ch in reversed(digits):
        n = int(ch)
        if alt:
            n *= 2
            if n > 9:
                n -= 9
        total += n
        alt = not alt
    return total % 10 == 0


def _key_is_sensitive(key: str, sensitive_keys: list[str]) -> bool:
    lower = key.lower()
    for needle in sensitive_keys:
        if needle in lower:
            return True
    return False


def _value_is_sensitive(
    value: str,
    patterns: list[Pattern[str]],
    scrub_emails: bool,
) -> bool:
    for pat in patterns:
        if pat.search(value):
            return True
    if scrub_emails and _EMAIL_RE.search(value):
        return True
    if _CC_RE.search(value) and _luhn_check(value):
        return True
    return False


def _scrub_value(
    value: Any,
    sensitive_keys: list[str],
    patterns: list[Pattern[str]],
    scrub_emails: bool,
    depth: int,
    seen: set[int],
) -> Any:
    if depth > 8:
        return value
    if value is None:
        return value
    if isinstance(value, str):
        return REDACTED if _value_is_sensitive(value, patterns, scrub_emails) else value
    if isinstance(value, (int, float, bool)):
        return value
    obj_id = id(value)
    if obj_id in seen:
        return value
    seen.add(obj_id)

    if isinstance(value, list):
        return [
            _scrub_value(v, sensitive_keys, patterns, scrub_emails, depth + 1, seen) for v in value
        ]

    if isinstance(value, dict):
        out: dict[str, Any] = {}
        for k, v in value.items():
            if isinstance(k, str) and _key_is_sensitive(k, sensitive_keys):
                out[k] = REDACTED
            else:
                out[k] = _scrub_value(v, sensitive_keys, patterns, scrub_emails, depth + 1, seen)
        return out

    return value


def scrub_event(
    event: dict[str, Any],
    options: Optional[dict[str, Any]] = None,
) -> dict[str, Any]:
    """Scrub PII from an event dict. Returns a new dict.

    Options:
        enabled (bool): Master switch, default True.
        sensitive_keys (list[str]): Extra key substrings to mask.
        scrub_emails (bool): Also mask email addresses, default False.
        custom_patterns (list[re.Pattern]): Extra value patterns.
    """
    if options and options.get("enabled") is False:
        return event

    opts = options or {}
    sensitive_keys = DEFAULT_SENSITIVE_KEYS + (opts.get("sensitive_keys") or [])
    sensitive_keys = [k.lower() for k in sensitive_keys]
    patterns = DEFAULT_VALUE_PATTERNS + (opts.get("custom_patterns") or [])
    scrub_emails = opts.get("scrub_emails", False)
    seen: set[int] = set()

    result = copy.deepcopy(event)

    for field in ("extra", "user", "tags"):
        if field in result and result[field]:
            result[field] = _scrub_value(
                result[field], sensitive_keys, patterns, scrub_emails, 0, seen
            )

    if "request" in result and result["request"]:
        req = result["request"]
        if "headers" in req and req["headers"]:
            req["headers"] = _scrub_value(
                req["headers"], sensitive_keys, patterns, scrub_emails, 0, seen
            )
        if "data" in req and req["data"]:
            req["data"] = _scrub_value(req["data"], sensitive_keys, patterns, scrub_emails, 0, seen)

    if "breadcrumbs" in result and result["breadcrumbs"]:
        for crumb in result["breadcrumbs"]:
            if "data" in crumb and crumb["data"]:
                crumb["data"] = _scrub_value(
                    crumb["data"], sensitive_keys, patterns, scrub_emails, 0, seen
                )

    return result
